backend/cn/lib/explain.ml

```diff
@@ -88,6 +89,53 @@ let subterms_without_bound_variables bindings =
     ITSet.empty
 
 
+(** Simplify a constraint in the context of a model. *)
+let simp_constraint eval lct =
+  let eval_to_bool it =
+    match eval it with Some (IT (Const (Bool b1), _, _)) -> Some b1 | _ -> None
+  in
+  let is b it = match eval_to_bool it with Some b1 -> Bool.equal b b1 | _ -> false in
+  let rec go (IT (term, bt, loc)) =
+    let mk x = IT (x, bt, loc) in
+    let ands xs = IT.and_ xs loc in
+    let go1 t = ands (go t) in
+    match term with
+    | Const (Bool true) -> []
+    | Binop (Or, lhs, rhs) when is false lhs -> go rhs
+    | Binop (Or, lhs, rhs) when is false rhs -> go lhs
+    | Binop (And, lhs, rhs) -> List.append (go lhs) (go rhs)
+    | Binop (Implies, lhs, rhs) ->
+      (match eval_to_bool lhs with
+       | Some b -> if b then go rhs else []
+       | None -> [ mk (Binop (Implies, go1 lhs, go1 rhs)) ])
+    | ITE (cond, ifT, ifF) ->
+      (match eval_to_bool cond with
+       | Some b -> if b then go ifT else go ifF
+       | None -> [ mk (ITE (go1 cond, go1 ifT, go1 ifF)) ])
+    | _ -> [ mk term ]
+  in
+  match lct with LC.Forall _ -> [ lct ] | LC.T ct -> List.map (fun x -> LC.T x) (go ct)
```

Comment:
We might want to extend `go` to know about binders (and not try to evaluate those), so we can apply the evaluation to `Forall` and to iterated resources.